File: src/main/java/dev/kikugie/elytratrims/client/resource/ETAtlasHolder.java
package dev.kikugie.elytratrims.client.resource;

import dev.kikugie.elytratrims.client.ETClient;
import dev.kikugie.elytratrims.client.config.TextureConfig;
import dev.kikugie.elytratrims.common.ETReference;
import net.minecraft.block.entity.BannerPattern;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.texture.*;
import net.minecraft.client.texture.atlas.AtlasLoader;
import net.minecraft.client.texture.atlas.Sprite;
import net.minecraft.registry.Registries;
import net.minecraft.registry.RegistryKey;
import net.minecraft.resource.ResourceFinder;
import net.minecraft.resource.ResourceManager;
import net.minecraft.resource.ResourceReloader;
import net.minecraft.util.Identifier;
import net.minecraft.util.profiler.Profiler;
import org.jetbrains.annotations.NotNull;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.function.Function;
import java.util.function.Supplier;

public class ETAtlasHolder implements ResourceReloader {
    /*? if >=1.20.2 */
    /*private static final SpriteOpener opener = SpriteOpener.create(SpriteLoader.METADATA_READERS);*/
    public static final Identifier TEXTURE = ETReference.id("textures/atlas/elytra_features.png");
    public static final Identifier ID = ETReference.id("elytra_features");
    private final SpriteAtlasTexture atlas;
    private boolean ready = false;

    public ETAtlasHolder() {
        atlas = new SpriteAtlasTexture(TEXTURE);
        MinecraftClient.getInstance().getTextureManager().registerTexture(ID, atlas);
    }

    public boolean isReady() {
        return ready;
    }

    public SpriteAtlasTexture getAtlas() {
        return atlas;
    }

    private List<Supplier<SpriteContents>> getSprites(ResourceManager manager) {
        Sprite elytra;
        try {
            elytra = ImageUtils.loadTexture(Textures.ELYTRA_MODEL, manager, 1);
        } catch (FileNotFoundException e) {
            ETReference.LOGGER.error("Failed to load elytra model texture");
            return List.of();
        }

        List<Supplier<SpriteContents>> sprites = new ArrayList<>(getTrims(manager, elytra));
        sprites.addAll(getPatterns(manager, elytra));
        sprites.add(getOverlay(elytra));
        sprites.add(MissingSprite::createSpriteContents);
        elytra.close();
        return sprites;
    }

    private Collection<Supplier<SpriteContents>> getTrims(ResourceManager manager, Sprite elytra) {
        var preSources = new AtlasLoader(ETResourceListener.getTrims()).loadSources(manager);
        /*? if <1.20.2 {*/
        var trimSources = preSources;
        /*?} else {*//*
        Collection<Supplier<SpriteContents>> trimSources = new ArrayList<>(preSources.size());
        for (Function<SpriteOpener, SpriteContents> func : preSources)
            trimSources.add(() -> func.apply(opener));
        *//*?} */
        return ETClient.getConfig().texture.cropTrims.get() ? ImageUtils.transform(trimSources, it -> ImageUtils.mask(it, elytra)) : trimSources;
    }

    private @NotNull Collection<Supplier<SpriteContents>> getPatterns(ResourceManager manager, Sprite elytraModel) {
        List<Supplier<SpriteContents>> patterns = new ArrayList<>();
        TextureConfig config = ETClient.getConfig().texture;
        ResourceFinder finder = new ResourceFinder("textures", ".png");
        for (RegistryKey<BannerPattern> key : Registries.BANNER_PATTERN.getKeys()) {
            patterns.add(() -> {
                Identifier id = BannerPattern.getSpriteId(key, config.useBannerTextures.get());
                Identifier texture = finder.toResourcePath(id);
                NativeImage pattern;
                try {
                    pattern = ImageUtils.loadTexture(texture, manager, 1).read();
                } catch (IOException e) {
                    ETReference.LOGGER.error("Failed to load pattern texture: {}", id);
                    return null;
                }

                pattern = config.useBannerTextures.get()
                        ? ImageUtils.dims(pattern, pattern.getWidth() * 2, pattern.getHeight())
                        : ImageUtils.dims(pattern, pattern.getWidth(), pattern.getHeight() / 2);
                int scale = pattern.getWidth() / 64;
                int xOffset = (int) ((config.useBannerTextures.get() ? 35.5F : 34F) * scale);
                int yOffset = config.useBannerTextures.get() ? (int) (scale * 1.5F) : 0;
                NativeImage offset = ImageUtils.offsetNotClosing(pattern, xOffset, yOffset, pattern.getWidth(), pattern.getHeight());

                return ImageUtils.mask(ImageUtils.createContents(offset, id), elytraModel);
            });
        }
        return patterns;
    }

    private Supplier<SpriteContents> getOverlay(Sprite elytraModel) {
        return () -> ImageUtils.createContents(ImageUtils.createSaturationMaskNotClosing(elytraModel), elytraModel.id.withPath(path -> path.replace("textures/", "").replace(".png", "")));
    }

    private CompletableFuture<List<SpriteContents>> transform(List<Supplier<SpriteContents>> sprites, Executor executor) {
        /*? if <1.20.2 {*/
        return SpriteLoader.loadAll(sprites, executor);
        /*?} else {*//*
        List<Function<SpriteOpener, SpriteContents>> transformed = new ArrayList<>(sprites.size());
        for (Supplier<SpriteContents> sup : sprites)
            transformed.add(o -> sup.get());
        return SpriteLoader.loadAll(opener, transformed, executor);
        *//*?} */
    }

    CompletableFuture<SpriteLoader.StitchResult> load(ResourceManager manager, Profiler ignoredProfiler, Executor executor) {
        return CompletableFuture
                .supplyAsync(() -> {
                    ready = false;
                    atlas.clear();
                    return getSprites(manager);
                }, executor)
                .thenCompose(sprites -> transform(sprites, executor))
                .thenApply(sprites -> SpriteLoader.fromAtlas(atlas).stitch(sprites, 0, executor))
                .thenCompose(SpriteLoader.StitchResult::whenComplete);
    }

    CompletableFuture<Void> apply(SpriteLoader.StitchResult data, ResourceManager ignoredManager, Profiler profiler, Executor executor) {
        return CompletableFuture.runAsync(() -> {
            profiler.startTick();
            profiler.push("upload");
            ETResourceListener.reset();
            ETClient.getRenderer().resetCache();
            atlas.upload(data);
            ready = true;
            profiler.pop();
            profiler.endTick();
        }, executor);
    }

    @Override
    public CompletableFuture<Void> reload(ResourceReloader.Synchronizer helper, ResourceManager manager, Profiler loadProfiler, Profiler applyProfiler, Executor loadExecutor, Executor applyExecutor) {
        return load(manager, loadProfiler, loadExecutor).thenCompose(helper::whenPrepared).thenCompose((o) -> apply(o, manager, applyProfiler, applyExecutor));
    }
}

File: src/main/kotlin/dev/kikugie/elytratrims/client/resource/ETAtlasHolder.kt
package dev.kikugie.elytratrims.client.resource

import com.google.gson.JsonParser
import com.mojang.blaze3d.systems.RenderSystem
import com.mojang.serialization.Dynamic
import com.mojang.serialization.JsonOps
import dev.kikugie.elytratrims.client.ETClient
import dev.kikugie.elytratrims.client.render.ETRenderer
import dev.kikugie.elytratrims.common.ETReference
import dev.kikugie.elytratrims.common.util.getAnyway
import dev.kikugie.elytratrims.common.util.identifier
import dev.kikugie.elytratrims.mixin.client.PalettedPermutationsAtlasSourceAccessor
import net.minecraft.client.MinecraftClient
import net.minecraft.client.texture.*
import net.minecraft.client.texture.SpriteLoader.StitchResult
import net.minecraft.client.texture.atlas.AtlasLoader
import net.minecraft.client.texture.atlas.AtlasSourceManager
import net.minecraft.client.texture.atlas.PalettedPermutationsAtlasSource
import net.minecraft.resource.ResourceFinder
import net.minecraft.resource.ResourceManager
import net.minecraft.resource.ResourceReloader
import net.minecraft.util.Identifier
import net.minecraft.util.profiler.Profiler
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor
import java.util.function.Function
import java.util.function.Supplier
import kotlin.jvm.optionals.getOrNull

object ETAtlasHolder : ResourceReloader {
    /*? if >=1.20.2*/
    /*private val opener: SpriteOpener = SpriteOpener.create(SpriteLoader.METADATA_READERS)*/
    val id: Identifier = ETReference.id("elytra_features")
    val atlas = SpriteAtlasTexture(ETReference.id("textures/atlas/elytra_features.png")).also {
        RenderSystem.recordRenderCall {
            @Suppress("UsePropertyAccessSyntax") // Neoforge sees it as a private property access idk why
            MinecraftClient.getInstance().getTextureManager().registerTexture(id, it)
        }
    }
    var ready = false
        private set

    private fun sprites(manager: ResourceManager, model: NativeImage, id: Identifier) = buildList {
        addAll(trims(manager, model))
        addAll(patterns(manager, model))
        add(color(id, model))
        addAll(animation(manager))
//        val item = loadTexture(identifier("textures/item/elytra.png"), manager)?.readSafe()
//        if (item != null) {
//            add(itemColor(ETReference.id("item/elytra_overlay"), item))
//            add(itemOutline(ETReference.id("item/elytra_outline"), item))
//            item.close()
//        }
        add(MissingSprite::createSpriteContents)
    }

    private fun patterns(manager: ResourceManager, model: NativeImage): Collection<ContentSupplier> {
        val useBanners = ETClient.config.texture.useBannerTextures
        val finder = ResourceFinder("textures/entity/${if (useBanners) "banner" else "shield"}", ".png")
        return buildList {
            for ((id, res) in finder.findResources(manager)) {
                var image = AtlasSprite(finder.toResourceId(id), res, 1).readSafe() ?: continue
                image = image.use {
                    if (useBanners) it.cropped(it.width * 2)
                    else it.cropped(height = it.height / 2)
                }
                val scale: Int = image.width / 64
                val xOffset = ((if (useBanners) 35.5F else 34F) * scale).toInt()
                val yOffset = if (useBanners) (scale * 1.5F).toInt() else 0
                add {
                    image
                        .use { it.offset(xOffset, yOffset) }
                        .use { it.mask(model) }
                        .toContents(id.withPath { path -> path.substring(9..<path.length - 4) })
                }
            }
        }
    }

    private fun trims(manager: ResourceManager, model: NativeImage): Collection<ContentSupplier> {
        val crop = ETClient.config.texture.cropTrims
        val atlases = manager.findAllResources("atlases") { it.path.endsWith("armor_trims.json") }
        val sources = atlases.values.flatten().flatMap {
            try {
                val dynamic = it.reader.use { Dynamic(JsonOps.INSTANCE, JsonParser.parseReader(it)) }
                val data = AtlasSourceManager.LIST_CODEC.parse(dynamic).getAnyway()
                data.filterIsInstance<PalettedPermutationsAtlasSource>()
            } catch (e: Exception) {
                emptyList()
            }
        }
        sources.forEach { src ->
            src as PalettedPermutationsAtlasSourceAccessor
            src.textures = src.textures
                .filter { "armor" in it.path && "leggings" !in it.path }
                .map { it.withPath { path -> path.replaceFirst("armor", "elytra") } }
        }

        fun SpriteContents.cropIfNeeded() = if (crop) transform { it.mask(model) } else this

        return AtlasLoader(sources).loadSources(manager).map {
            {
                val sprite = /*? if <1.20.2 {*/it.get()/*?} else*//*it.apply(opener)*/
                sprite?.cropIfNeeded()
            }
        }
    }

    private fun color(id: Identifier, model: NativeImage): ContentSupplier = { saturationMask(model).toContents(id) }

    private fun itemOutline(id: Identifier, model: NativeImage): ContentSupplier {
        val out = model.outline { edge -> if (edge) -1 else 0 }
        return { out.toContents(id) }
    }

    private fun itemColor(id: Identifier, model: NativeImage): ContentSupplier {
        val out = saturationMask(model).outline { edge -> if (edge) 0 else null }
        return { out.toContents(id) }
    }

    private fun animation(manager: ResourceManager): Collection<ContentSupplier> = buildList {
        val resource = manager.getResource(ETReference.id("textures/animation/animation.png")).getOrNull()
            ?: return emptyList()
        /*? if <1.20.2 {*/
        val sprite = SpriteLoader.load(ETReference.id("animation/animation"), resource) ?: return emptyList()
        /*?} else*/
        /*val sprite = opener.loadSprite(ETReference.id("animation/animation"), resource) ?: return emptyList()*/
        add { sprite }
    }

    private fun transform(
        sprites: List<ContentSupplier>,
        executor: Executor,
    ): CompletableFuture<List<SpriteContents>> =
        /*? if <1.20.2 {*/SpriteLoader.loadAll(sprites.mapNotNull(::asSupplier), executor);
        /*?} else*//*SpriteLoader.loadAll(opener, sprites.mapNotNull(::asFunction), executor)*/

    private fun load(manager: ResourceManager, executor: Executor): CompletableFuture<StitchResult> {
        var model: NativeImage? = null
        return CompletableFuture.supplyAsync {
            ready = false
            ETRenderer.reset()
            atlas.clear()
            val id = identifier("textures/entity/elytra.png")
            model = loadTexture(id, manager)?.readSafe() ?: return@supplyAsync emptyList()
            sprites(manager, model!!, id.withPath { path: String ->
                path.replace("textures/", "").replace(".png", "")
            })
        }.thenCompose {
            transform(it, executor)
        }.thenApply {
            SpriteLoader.fromAtlas(atlas).stitch(it, 0, executor)
        }.thenCompose(StitchResult::whenComplete)
            .whenComplete { _, _ -> model?.close() }
    }

    fun apply(
        data: StitchResult,
        profiler: Profiler,
        executor: Executor,
    ): CompletableFuture<Void> = CompletableFuture.runAsync({
        profiler.startTick()
        profiler.push("upload")
        atlas.upload(data)
        ready = true
        profiler.pop()
        profiler.endTick()
    }, executor)

    override fun reload(
        helper: ResourceReloader.Synchronizer,
        manager: ResourceManager,
        loadProfiler: Profiler,
        applyProfiler: Profiler,
        loadExecutor: Executor,
        applyExecutor: Executor,
    ): CompletableFuture<Void> = load(manager, loadExecutor)
        .thenCompose {
            helper.whenPrepared(it)
        }.thenCompose {
            apply(it, applyProfiler, applyExecutor)
        }

    internal fun <T> asSupplier(it: (() -> T)?) = if (it != null) Supplier { it() } else null
    private fun <P, T> asFunction(it: (() -> T)?) = if (it != null) Function<P, T> { it() } else null
}